Add kubevirt_node_deprecated_machine_types metric

dasionov opened the pull request:
### What this PR does

This PR introduces two changes:

- Refactors the node-labeller controller to accept `[]libvirtxml.CapGuestMachine` as an argument instead of `[]libvirtxml.CapGuest`.

- Adds a new dedicated metric to expose the list of deprecated machine types per each node in the KubeVirt cluster.

Before this PR:
KubeVirt did not have a dedicated metric to expose information about the deprecated machine types per each node in the cluster.

After this PR:
A dedicated metric is now available, listing all deprecated machine types per node that can be used for alerting purposes.

## Query Screenshot:
![image](https://github.com/user-attachments/assets/df700c78-9061-4e2c-ab5f-7b8fed70e07e)
Fixes #

### Why we need it and why it was done in this way
This lays the groundwork for future alerting features. It will help detect when virtual machines are provisioned using outdated or deprecated machine types that require upgrades.

### Release note
```release-note
None
```

dasionov commented:
/cc @enp0s3
/cc @machadovilaca

machadovilaca commented:
I wonder if a collector approach, that used the labels from the nodes, and exported the metrics from there, would be a cleaner approach. wdyt? @dasionov

dasionov commented:
> I wonder if a collector approach, that used the labels from the nodes, and exported the metrics from there, would be a cleaner approach. wdyt? @dasionov

I’ve been thinking about this as well, and I IMO a collector approach seems more suited for data that is dynamic and dependent on other components being provisioned.
A good example of this would be the domainstat metrics, which aren’t populated until a VMI is actually created.

In contrast, the supported machine types are more static—they only change when a node is upgraded, and even then, typically just once per upgrade cycle. Given this, it feels like the collector approach might be less ideal here, as it introduces additional overhead for something that doesn’t change frequently.

iholder101 commented:
I believe the first commit is more relevant to sig-compute while the second one to sig-observability.

From the sig-compute side, this looks great. Thanks @dasionov.
You're welcome to ping me for re-approval (if needed) after this is approved by observability.

/lgtm

avlitman commented on tests/monitoring/metrics.go:
```diff
@@ -106,7 +106,7 @@ var _ = Describe("[sig-monitoring]Metrics", decorators.SigMonitoring, func() {
 			Expect(err).ToNot(HaveOccurred())
 
-			err = virthandler.SetupMetrics("", "", 0, nil)
+			err = virthandler.SetupMetrics("", "", 0, nil, nil)
```
the metric is not created once it's nil

dasionov replied:
i added the metric name to the map of excluded matrics, since we already testing the metric in `tests/infrastructure/prometheus`

avlitman commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -48,15 +48,15 @@ var _ = Describe("Node-labeller ", func() {
 	var nlController *NodeLabeller
```
please add unit-tests

avlitman replied:
yes

dasionov commented:
/test pull-kubevirt-build-s390x

dasionov commented:
@avlitman the test in [pull-kubevirt-e2e-k8s-1.31-sig-monitoring](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/14255/pull-kubevirt-e2e-k8s-1.31-sig-monitoring/1902030051533131776)
fails on this metric `kubevirt_rest_client_rate_limiter_duration_seconds`
seems unrelated to my changes.

dasionov commented:
@machadovilaca
My issue with the current implementation is that I have duplicated entries because there are labels for the handler instance in addition to the machine_type and the deprecated labels I need.

machadovilaca commented:
> @machadovilaca My issue with the current implementation is that I have duplicated entries because there are labels for the handler instance in addition to the machine_type and the deprecated labels I need.

You didn't want to have the metric per node? Although each node may support different machine types,
you wanted to aggregate all the machines supported in the cluster, is that correct?


In this approach, each virt-handler will expose its own metrics, hence the duplication, but I think that should be fine. In my opinion, you can just add a node label with the node name to the metric

dasionov commented:
> > @machadovilaca My issue with the current implementation is that I have duplicated entries because there are labels for the handler instance in addition to the machine_type and the deprecated labels I need.
> 
> You didn't want to have the metric per node? Although each node may support different machine types, you wanted to aggregate all the machines supported in the cluster, is that correct?
> 
> In this approach, each virt-handler will expose its own metrics, hence the duplication, but I think that should be fine. In my opinion, you can just add a node label with the node name to the metric

I thought of it too, but i don't see how it help reduce the duplication.. if you are saying it can't be helped then i guess we will have to live with it

machadovilaca commented:
in virt-handler its tricky because we don't have a "leader"
the node suggestion was to accept that duplication

but one thing that I don't understand yet is whether the machines supported are per node or per cluster?

dasionov commented:
> in virt-handler its tricky because we don't have a "leader" the node suggestion was to accept that duplication
> 
> but one thing that I don't understand yet is whether the machines supported are per node or per cluster?

each node, reports the list of supported machine types available for VMs to use so they can be scheduled on that specific node.

machadovilaca commented:
Then I think it's fine for each virt-handler to report the machines supported in their current node with the mentioned "node" label

dasionov commented:
> Then I think it's fine for each virt-handler to report the machines supported in their current node with the mentioned "node" label

added the node label as well PTAL :pray:

dasionov commented:
/test pull-kubevirt-unit-test-s390x

machadovilaca commented:
/approve on the monitoring side

@dasionov update the image on the description

0xFelix commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -300,13 +300,18 @@ func (app *virtHandlerApp) Run() {
 		panic(err)
 	}
 
+	var supportedMachines []libvirtxml.CapsGuestMachine
+	for _, guest := range capabilities.Guests {
+		supportedMachines = append(supportedMachines, guest.Arch.Machines...)
+	}
```
nit: keep that in a separate helper?

dasionov replied:
done

0xFelix commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -300,13 +300,18 @@ func (app *virtHandlerApp) Run() {
 		nodeLabellerrecorder,
 		capabilities.Host.CPU.Counter,
-		capabilities.Guests,
+		supportedMachines,
```
Could we need additional information in `capabilities.Guests` in the future?

dasionov replied:
seems unlikely IMO

0xFelix commented on tests/infrastructure/prometheus.go:
```diff
@@ -419,6 +419,25 @@ var _ = Describe(SIGSerial("[rfe_id:3187][crit:medium][vendor:cnv-qe@redhat.com]
 	)
 
+	DescribeTable("should expose kubevirt_supported_machine_types metric with supported types", func(family k8sv1.IPFamily) {
+		libnet.SkipWhenClusterNotSupportIPFamily(family)
```
How do supported machine types relate to an ip address family?

dasionov replied:
it's not related, the metrics extraction in this file and similar tests here, being done through command inside the virt-handler, and usually there are 2 tests for that, one with ipv4, and the other with ipv6
i could just stay with the ipv4 default one, if that's ok

0xFelix replied:
I have to admit that I find it rather strange that many of these tests differentiate between IPv4 and IPv6 at all.

enp0s3 replied:
@0xFelix Hi, I share your concern but I think the path forward if to open an issue and work on it separately. For the current PR I would recommend to stick with the strange pattern for the sake of test-infra deps

dasionov commented:
/retest-required

dasionov commented:
/test pull-kubevirt-e2e-k8s-1.32-sig-compute-serial

avlitman commented:
/lgtm

dasionov commented:
> Hi @dasionov , How do you mean to use the metric? I want to make sure current implementation will support it well.

See the alert I want to be based on this:
https://github.com/kubevirt/hyperconverged-cluster-operator/pull/3358

> 
> I see you also report deprecated machine_types. How will they be used, why not report only the supported machine_types (We usually report only what is supported)?

I was thinking about this too, with @enp0s3 we reached to a conclusion that during upgrades new virt-handlers will report an old metric as deprecated (so the supported ones are those without a deprecated label on it.)
Thus, the deprecated label is important for the alert

sradco commented:
> I see you also report deprecated machine_types. How will they be used, why not report only the supported machine_types (We usually report only what is supported)?
> 
> 

> > I was thinking about this too, with @enp0s3 we reached to a conclusion that during upgrades new virt-handlers will report an old metric as deprecated (so the supported ones are those without a deprecated label on it.)
> > Thus, the deprecated label is important for the alert

I think, based on the alert you plan to create, that it might be better to only report the deprecated machine types.
Commented in the alert as well

dasionov commented:
> I think, based on the alert you plan to create, that it might be better to only report the deprecated machine types. Commented in the alert as well

I agree, fixed the metric to contain only deprecated machine-types, also dropped the node-name seems irrelevant IMO
will update the image in the description

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/71497c3428705ac29c5a5d5af49cce1265dafafa..9b4bab0e1311d0abc692cfb71ec65fd0693b9dca): Address @sradco's [Comment](https://github.com/kubevirt/kubevirt/pull/14255#issuecomment-2880293741)

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/9b4bab0e1311d0abc692cfb71ec65fd0693b9dca..bdcef6a507f7228e13f14ad1773fd8cd96153cb2): Add licence to `machine-type.go`, `machine_type_test.go`

enp0s3 commented:
/test pull-kubevirt-code-lint

enp0s3 commented on pkg/monitoring/metrics/virt-handler/machine_type.go:
```diff
@@ -0,0 +1,44 @@
+/*
+Copyright The KubeVirt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
```suggestion
/*
 * This file is part of the KubeVirt project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Copyright The KubeVirt Authors.
 *
 */
```

dasionov replied:
fixed

enp0s3 commented on pkg/monitoring/metrics/virt-handler/machine_type_test.go:
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright The KubeVirt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
```
```suggestion
/*
 * This file is part of the KubeVirt project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Copyright The KubeVirt Authors.
 *
 */
```

dasionov replied:
fixed

enp0s3 reviewed (commented):
@dasionov Thank you for the effort!
A few arrangements with the commits, other than that LGTM.

Can you please couple the implementation of `ReportDeprecatedMachineTypes` and its unit testing in one commit?

Also can you please squash the rework you did on the the new e2e tests you've added in `tests/infrastructure/prometheus.go` into one commit?

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/e03bb86b92fc9884b08cdffbb65cab3a9dcc3d1a..0a0a68500d54d399310be6ab2eeeb57285b2ded4): Squash commits, and fix licencing

dasionov commented:
/test pull-kubevirt-unit-test-s390x

sradco commented:
@machadovilaca @avlitman I see that the metric is reported twice for each virt-handler. I believe there is a way to avoid this by using the leader one. Right? Can you please let @dasionov know what he should do?

dasionov commented:
> @machadovilaca @avlitman I see that the metric is reported twice for each virt-handler. I believe there is a way to avoid this by using the leader one. Right? Can you please let @dasionov know what he should do?

I don't mind the duplication that much,  is the rest of the metric and changes looks ok ?
@enp0s3 @sradco

sradco commented:
We shouldn't report it twice..

sradco commented:
@machadovilaca please see if you can help here.

machadovilaca commented:
> @machadovilaca @avlitman I see that the metric is reported twice for each virt-handler. I believe there is a way to avoid this by using the leader one. Right? Can you please let @dasionov know what he should do?

@sradco @dasionov 

virt-handler does not have a "leader" as each one is responsible for the node where its running
with that, this metric is implemented in virt-handler, because it may differ from node to node, isn't that correct?

if so, then it's okay to have the repeated values for the same machine type, in my opinion, as they are specific to that node and not global

dasionov commented:
> @machadovilaca But from my understanding this is not a per node metric an should show a cluster level metric.

Actually, each virt-handler reports the machine types available on its own node. In most clusters, nodes have the same capabilities, which leads to some duplication. But overall, the metric is reported by each virt-handler based on the capabilities of its specific node—so it’s actually a per-node metric.

I dropped the `nodeName` label because I didn’t find it particularly useful for this metric right now. We can always add it later if we find a reason to, or if it’s considered standard practice.

During cluster upgrades, Some nodes might run newer OS versions earlier, which could deprecate certain machine types before others. That would create temporary differences between nodes until the upgrade is complete.

Another scenario is clusters made up of completely different types of nodes, like in multi-arch environments, where machine type support can vary more significantly.

machadovilaca commented:
> > @machadovilaca But from my understanding this is not a per node metric an should show a cluster level metric.
> 
> Actually, each virt-handler reports the machine types available on its own node. In most clusters, nodes have the same capabilities, which leads to some duplication. But overall, the metric is reported by each virt-handler based on the capabilities of its specific node—so it’s actually a per-node metric.
> 
> I dropped the `nodeName` label because I didn’t find it particularly useful for this metric right now. We can always add it later if we find a reason to, or if it’s considered standard practice.
> 
> During cluster upgrades, Some nodes might run newer OS versions earlier, which could deprecate certain machine types before others. That would create temporary differences between nodes until the upgrade is complete.
> 
> Another scenario is clusters made up of completely different types of nodes, like in multi-arch environments, where machine type support can vary more significantly.

@dasionov since this caused confusion here, same will happen for users, I would suggest making in explicit in the metric name and in the help that these are per node

sradco commented:
@dasionov please update pr summary and description

dasionov commented:
> @dasionov please update pr summary and description

PTAL :pray:

dasionov commented:
@enp0s3 does everything looks OK to you?

enp0s3 commented:
@Barakmor1 Hi, thanks for bringing it up. I think that in general when we have large scale cluster the Prometheus should be prepared to scrape large amount of node metric data. IIUC the load will be on the Prometheus side, thus virt-handler isn't under risk of stability.
Can you estimate the size of he deprecated machines list?

sradco commented:
That is correct. We do not limit the number of metrics and Prometheus is built to scrape large amounts of metrics. This is not a concern.

dasionov commented:
/retest-required

dasionov commented:
[Change](https://github.com/kubevirt/kubevirt/compare/2fecb778f069780b4145f74c0cbbced875c429a1..628619b6f5fcecc9dfee66e4ffb670263d91fa7d): Rebase and updated import of `operator-observability` to `/rhobs/operator-observability-toolkit/`

Barakmor1 commented:
> @Barakmor1 Hi, thanks for bringing it up. I think that in general when we have large scale cluster the Prometheus should be prepared to scrape large amount of node metric data. IIUC the load will be on the Prometheus side, thus virt-handler isn't under risk of stability. 

Yes, my main concern was about potentially overloading Prometheus with too many metrics. But as you mentioned, it’s probably designed to handle large-scale metric workloads. The only reference I found that touches on something similar is from the Prometheus docs: https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels.

> Can you estimate the size of he deprecated machines list?

As far as I know, there are usually around 10 deprecated machine types, though I'm not completely sure. Anyway, I thought it was worth mentioning.
